allocator.go

```diff
@@ -184,3 +223,40 @@ func (a *Allocator) Allocate() (ok bool, assignments map[RangeID][]NodeID) {
 	}
 	return true, res
 }
+
+// Minimize the number of Assignments which were 1 but would now become 0
+// Note: If a Range or a Node is no longer available and that causes an existing allocation to be removed,
+// it will not count as a churn
+func (a *Allocator) adhereToChurnConstraint() {
+
+	var toMinimizeTheSumLiterals = make([]cpsatsolver.Literal, 0, len(a.ranges)*len(a.nodes))
+	prevAssignmentMap := a.opts.churnOptions.prevAssignment
+
+	for _, r := range a.ranges {
+		for _, n := range a.nodes {
+			if previousAssignment, ok := prevAssignmentMap[r.id][n.id]; ok == true && previousAssignment == 1 {
+				// Literal below would exist in the map because we are iterating through the new ranges and nodes
+				newLiteral := a.literal(r, n)
+				// previousAssignment = 1 and newLiteral.Not() would be true if the new value is 0. Minimize the sum of these
+				toMinimizeTheSumLiterals = append(toMinimizeTheSumLiterals, newLiteral.Not())
+			}
+		}
+	}
+
+	if a.opts.churnOptions.withMinimalChurn {
+		a.model.Minimize(cpsatsolver.Sum(asIntVars(toMinimizeTheSumLiterals)...))
+	}
+	if a.opts.churnOptions.withMaxChurn {
+		a.model.AddConstraints(
+			cpsatsolver.NewAtMostKConstraint(int(a.opts.churnOptions.maxChurn), toMinimizeTheSumLiterals...),
+		)
+	}
+}
+
+func asIntVars(literals []cpsatsolver.Literal) []cpsatsolver.IntVar {
+	var res []cpsatsolver.IntVar
+	for _, l := range literals {
+		res = append(res, l.(cpsatsolver.IntVar))
+	}
+	return res
+}
```

Comment:
You can declare `toMinimizeTheSumLiterals` as an array of `IntVar`, and then that would eliminate the need for this method, `Literal` and `IntVar` both implement the same interface. Although you'd then need to use `LinearConstraint` as opposed to `NewAtMostKConstraint`.

Reply:
The way it is now might not be as optimized but is more readable don't you think?

Reply:
By the way, the pattern` Sum(asIntVars(literals)...)` is used multiple times even in the Solver library at `constraint.go`. Perhaps `asIntVars` needs to be a public function.

Any opinion on this @irfansharif

Reply:
Yeah both implementations are fine by me. 👍

Reply:
It's public: https://godocs.io/github.com/irfansharif/solver#AsIntVars

Reply:
Not public on the version we are using in this PR, will update to use it accordingly afterwards.

Reply:
I'll fix this once I put my PR up since I bumped up deps, it's fine to merge this as is ATM.